scripts/uploaded-files/trash-legacy-files-in-s3.ts

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env ./node_modules/.bin/babel-node
+
+/**
+ * This script assumes that all files still being used have been recorded to the `UploadedFile` table (using `scripts/uploaded-files/record-existing-files.ts`).
+ * It will move all files that are in S3 but not in the database to a trash folder by prepending `trash/` to the key.
+ */
+
+import '../../server/env';
+
+import fs from 'fs';
+
+import { S3 } from 'aws-sdk';
+import { Command } from 'commander';
+import config from 'config';
+import PQueue from 'p-queue';
+
+import { listFilesInS3, S3_TRASH_PREFIX, trashFileFromS3 } from '../../server/lib/awsS3';
+import logger from '../../server/lib/logger';
+import { sequelize } from '../../server/models';
+
+const DRY_RUN = process.env.DRY_RUN !== 'false';
+const LOCAL_CACHE_FILE = 'output/s3-files.json';
+
+const getNonTrashedFilesFromS3 = async () => {
+  let objects: S3.Object[] = [];
+  let message;
+
+  if (fs.existsSync(LOCAL_CACHE_FILE)) {
+    // Load the list from local file if it exists
+    objects = JSON.parse(fs.readFileSync(LOCAL_CACHE_FILE, 'utf8'));
+    message = `Found ${objects.length} files in local cache`;
+  } else {
+    // Otherwise fetch all files from S3 and save the list locally (in case we need to restart the script)
+    objects = await listFilesInS3(config.aws.s3.bucket);
+    message = `Found ${objects.length} files in S3. Saving list to local cache`;
+    fs.writeFileSync(LOCAL_CACHE_FILE, JSON.stringify(objects, null, 2));
+  }
+
+  const nonTrashedObjects = objects.filter(o => !o.Key.startsWith(S3_TRASH_PREFIX));
+  logger.info(`${message} (${nonTrashedObjects.length} non-trashed files)`);
+  return objects;
+};
+
+const main = async options => {
+  const concurrency = parseInt(options['concurrency']) || 3;
+  const allObjects = await getNonTrashedFilesFromS3();
```

Comment:
Around how many objects are we handling here? Does it work fine as an input to the sql query below?

Reply:
~240.000 objects. It does work fine, with the query below, mainly because we're taking advantage of the index by doing a strict equal comparison.